pkg/window/strategy/session/session.go

```diff
@@ -0,0 +1,389 @@
+	// with any existing window, WindowToBeMerged returns the window that can be merged
+	// if the returned window is not same as the window created by the message, that means
+	// we need to expand the window to accommodate the new message and send an expand operation
+	// to the server
+	if windowToBeMerged, canBeMerged := list.WindowToBeMerged(win); canBeMerged {
+		// check whether we have to do expand window.
+		if win.StartTime().Before(windowToBeMerged.StartTime()) || win.EndTime().After(windowToBeMerged.EndTime()) {
+			// we need to clone the window because for expand operation we need to send the old window (e.g., 60-70)
+			// and the expanded window (e.g., 60-72).
+			oldWindow := cloneWindow(windowToBeMerged)
+
+			// update the reference of the window in the list with the expanded window
+			windowToBeMerged.Merge(win)
+			windowOperations = append(windowOperations, createWindowOperation(message, window.Expand, []window.TimedWindow{oldWindow, windowToBeMerged}, &Partition))
+		} else { // no need to expand, message doesn't change the window end-time.
+			// if the returned window is same as the window created by the message, that means
+			// the message belongs to an existing window and it's an append operation
+			windowOperations = append(windowOperations, createWindowOperation(message, window.Append, []window.TimedWindow{windowToBeMerged}, &Partition))
+		}
+
+		return windowOperations
+	}
+
+	// if the window cannot be merged, that means we need to create a new window to the active windows list
+	list.Insert(win)
+	windowOperations = append(windowOperations, createWindowOperation(message, window.Open, []window.TimedWindow{win}, &Partition))
+
+	return windowOperations
+}
+
+// InsertWindow inserts window to the list of active windows
+func (w *Windower) InsertWindow(tw window.TimedWindow) {
+	combinedKey := strings.Join(tw.Keys(), delimiter)
+	if list, ok := w.activeWindows[combinedKey]; !ok {
+		list = window.NewSortedWindowListByEndTime()
+		list.InsertFront(tw)
+		w.activeWindows[combinedKey] = list
+	} else {
+		list.Insert(tw)
+	}
+}
+
+func createWindowOperation(message *isb.ReadMessage, event window.Operation, windows []window.TimedWindow, id *partition.ID) *window.TimedWindowRequest {
+	// clone the windows because the windows might be updated after the operation is sent to the server.
+	// we do inplace updates, but those will be merged later on, hence correctness won't be affected.
+	var clonedWindows = make([]window.TimedWindow, 0)
+	for _, win := range windows {
+		clonedWindows = append(clonedWindows, cloneWindow(win))
+	}
+	return &window.TimedWindowRequest{
+		ReadMessage: message,
+		Operation:   event,
+		Windows:     clonedWindows,
+		ID:          id,
+	}
+}
+
+// CloseWindows closes the windows that are past the watermark
+// and also merges the windows that should be merged.
+func (w *Windower) CloseWindows(time time.Time) []*window.TimedWindowRequest {
```

Comment:
`CloseAndMergeWindows`?

Reply:
I prefer `CloseWindows`, since its a common interface for both aligned and unaligned windows